crates/maple_cli/src/cmd/filter.rs

```diff
@@ -70,3 +70,409 @@ pub fn blines(
+        // the queue with best results the same size.
+        let mut past = std::time::Instant::now();
+        iter.for_each(|(text, score, indices)| {
+                // Best results are stored in front.
+                //XXX I can't say, if bigger score is better or not. Let's assume the bigger the better.
+                let idx = top_scores
+                    .iter()
+                    .rev() // .rev(), because worse items are at the end.
+                    .enumerate()
+                    .find(|&(_, &other_score)| other_score > score);
+                insert_both!(pop; idx.map(|(i, _)| i), score, text, indices => buffer, top_results, top_scores);
+
+                total += 1;
+                if total % 16 == 0 {
+                    use std::time::{Duration, Instant};
+
+                    const UPDATE_INTERVAL: Duration = Duration::from_secs(2);
+
+                    let now = Instant::now();
+                    if now > past + UPDATE_INTERVAL {
+                        past = now;
+                        println_json!(total);
+                        for &idx in top_results.iter() {
+                            let (text, _, indices) = std::ops::Index::index( buffer.as_slice(), idx);
+                            println_json!(text, indices);
+                        }
+                    }
+                }
+
+                if buffer.len() == buffer.capacity() {
+                    buffer.par_sort_unstable_by(|(_, v1, _), (_, v2, _)| v2.partial_cmp(&v1).unwrap());
+
+                    for (idx, (_, score, _)) in buffer[..ITEMS_TO_SHOW].iter().enumerate() {
+                        top_scores[idx] = *score;
+                        top_results[idx] = idx;
+                    }
+
+                    let half = buffer.len() / 2;
+                    buffer.truncate(half);
+                }
+            });
+
+        (total, buffer)
+    }
+
+    if let Some(number) = number {
+        let (total, filtered) = match source {
+            Source::Stdin => dyn_collect_number(
+                io::stdin().lock().lines().filter_map(|lines_iter| {
+                    lines_iter.ok().and_then(|line| {
+                        scorer(&line).map(|(score, indices)| (line, score, indices))
+                    })
+                }),
+                number,
+            ),
+            Source::File(fpath) => dyn_collect_number(
+                std::fs::read_to_string(fpath)?.lines().filter_map(|line| {
+                    scorer(&line).map(|(score, indices)| (line.into(), score, indices))
+                }),
+                number,
+            ),
+            Source::List(list) => dyn_collect_number(
+                list.filter_map(|line| {
+                    scorer(&line).map(|(score, indices)| (line, score, indices))
+                }),
+                number,
+            ),
+        };
+
+        let ranked = filtered;
+
+        let payload = ranked.into_iter().take(number);
```

Comment:
I see some code repeations here. I can refactor it later if you are focusing the dyn part.

Reply:
I was just putting all things into one function to rewrite it as dynamic, and this renaming shows a borderline between old functions.

And I don't think I'll be doing any other big changes to this particular function, because now it has no unsafe and works. But, for example, if you want to search in all files in a folder and its subfolders *in parallel*, the filtering and dynamic updates should be done other way.

Reply:
>if you want to search in all files in a folder and its subfolders effective and in parallel, the filtering and dynamic updates should be done other way.

clap iterself does not do the job of searching, but relays on the tools like `fd`, `ripgrep`, which already does the best in terms of searching. What clap mainly does is to do the filtering on the data of `fd`/`ripgrep` produces, so we don't have to worried about this point.

Reply:
> 
> 
> > if you want to search in all files in a folder and its subfolders effective and in parallel, the filtering and dynamic updates should be done other way.
> 
> clap iterself does not do the job of searching, but relys on the tools like `fd`, `ripgrep`, which already does the best in terms of searching. What clap mainly does is to do the filtering on the data of `fd`/`ripgrep` produces, so we don't have to worried about this point.

So, you make a lot of serializing-deserializing between fd / ripgrep and python module, and then again ser-deser from python to maple. Or am I wrong?

Reply:
Nope. Actually the Python part is another topic. The Python dynamic module is only used when all the data is ready and for the small dataset as:

1. The Python filtering  is sync which can block vim if there are too much data.
2. There is no need to spawn a `job` in vim to communiate with the external binary maple.

When the dataset is huge, e.g., 1 million lines, I just use the `std::process::Command` to execute the fd/ripgrep command, gather their output and then do the filtering. I know it's not good enought, but at least vim won't be blocked as all the heavy job runs externally.

So my final goal for maple is to run the job like `fd`/`ripgrep` for producing data in one thread, reading the output in another thread and do the filtering, which is mostly what current popular fuzzy finders like `fzf`/`skim` do. The difference between `fzf`/`skim` and vim-clap is that `fzf`/`skim` are essentially TUI programs even they can also used in vim, drawing the UI by theirselves, for vim-clap, it uses the native Vim UI stuffs.

Reply:
> I just use the `std::process::Command` to execute the fd/ripgrep command, gather their output and then do the filtering

And what is the output? Is it files' handles, lines, or something else?

To not talk about this unfounded, that's how I see effective fuzzy-filtering of many files with dynamic results:

1. There's always a number-cap for results. If it's not provided by the user, it's somethings like 500, 1000 or 2000, some number of results that is hard to read for humans but easy to store for computers (and if that's not enough for the user, he can provide a bigger number and run filtering again). This is almost always better than absence of such cap, and always better for big numbers of total results.
2. Bigger tasks are parallel, lesser aren't. Instead of parallel reads of lines in one file, which is not very effective, threads share the pool of files or lines and take some files or many lines at once, then process those in old one-threaded manner, synchronizing results from time to time (how and when to synchronize is a lower-level details, that aren't clear just now). Now, instead of many short-living threads, that fight for OS resources each time they are created, we have some long-living threads, that are almost always working.

Simply saying, all this optimizing is around less number of reallocations and more effective use of threads.

Reply:
> And what is the output? Is it files' handles, lines, or something else?

Just the `Output` of `std::process::Command`, see https://github.com/liuchengxu/vim-clap/blob/master/crates/maple_cli/src/light_command.rs#L78-L90. I'll then convert this `Output` to lines for the filtering.

It's ok to always enforce a number-cap as that's the typical use case, and the screen size is limited, people normally won't check out more than dozens of items.

Reply:
@ImmemorConsultrixContrarie Yes, one vim instanace has only one maple-bin running, but the maple process can be created and killed multiple times per vim instance context.

Reply:
> Just the `Output` of `std::process::Command`, see https://github.com/liuchengxu/vim-clap/blob/master/crates/maple_cli/src/light_command.rs#L78-L90. I'll then convert this `Output` to lines for the filtering.

Yeah, I see how you get `&[u8]` from the output, but can't find where it's passed into fuzzy-filter, only more `println!`s.

Reply:
hmm, it's a little complex. Due to the missing of dynamic update, gathering the output of `Command` and filtering given the query is splitted into two steps with some temp naive optimization, such as writting to a temp file. But that would be unnecccary with the dyamic update, now we can just feed `&[u8]` into the filter.

Reply:
So, do we change the current filterer, so it works with `&[u8]` and always-capped number, and then land this filterer, even though it's one-threaded right now; or do we write a proper multithreaded filterer, that is focused on working with `&[u8]` and always-capped number?

Or even more epic task: merging ripgrep / fd code with fuzzy-filter, so whenever you find a line, you fuzzy-filter it in place and either store it or throw away. This is better for both thread-usage and RAM problems like #292, as there will always be a cap number and no more than that number of lines will be ever stored.

Reply:
By the way, rg/fd search for regex, but fuzzy-filter isn't regexed, so linefinder for fuzzy-filter could be very simple. Or am I wrong and you can fuzzy-search regexes?

Reply:
>merging ripgrep / fd code with fuzzy-filter,

That's crazy 🙉 , that could be hard to maintain I think.

I think it's a good choice to write a multithreaded filter, one thread A for producing data given the fd/ripgrep command, the other for filtering on the data produced by thread A and notifying the result periodly.

Reply:
fuzzy-search regexes is unsupported at the moment, but it sounds like a great addition!  @ImmemorConsultrixContrarie

Reply:
> 
> 
> fuzzy-search regexes is unsupported at the moment, but it sounds like a great addition! @ImmemorConsultrixContrarie

I can't even imagine how this thing could be implemented (apart from very slow implementation, when you make a regex into all strings it could represent and then fuzzy-search for each in each line) . Now THIS is crazy, man.

So, right now you can search for regex OR fuzzy-search some letters in line. And the latter task doesn't need any regex search, actually, just some uses of [memchr](https://docs.rs/memchr/2.3.3/memchr/). And I can merge such a simple linefinder with fuzzy-filterer for sure. And [ignore](https://docs.rs/ignore/0.4.13/ignore/) crate will help with recursive directory search.

So, if you don't plan writing regex fuzzy-finder yourself, I'll just stick to this very simple plan, that will optimize RAM usage, thread usage and probably make fuzzy-search simply faster.

Reply:
https://github.com/junegunn/fzf#search-syntax Although fzf does not support full regexp based searhcing, it has a extended search mode, and I think it's useful, I might also want to introduce it in clap in the long term.